scripts/Muon/GUI/ElementalAnalysis2/context/ea_group_context.py

```diff
@@ -0,0 +1,109 @@
+    '''this creates the first set of groups listed in the grouping table of the Elemental Analysis GUI '''
+    groups = []
+    run_list = loadedData.get_parameter("run")
+    for run_item in run_list:
+        for workspace in loadedData.get_data(run=run_item)["workspace"]:
+            group_name = str(workspace)
+            '''For single workspace names the detector is found by taking everything after ; in the name
+            For co-added workspaces the detector is found by taking everything before '''
+            detector_name = (group_name.split(';', 1)[-1].lstrip()).split('_', 1)[0]
+            run_number = str(run_item).replace('[', '').replace(']', '')
+            groups += [EAGroup(group_name=group_name, detector=detector_name, run_number=run_number)]
+    return groups
+
+
+class MessageNotifier(Observable):
+    def __init__(self, outer):
+        Observable.__init__(self)
+        self.outer = outer  # handle to containing class
+
+    def notify_subscribers(self, *args, **kwargs):
+        Observable.notify_subscribers(self, *args)
+
+
+class EAGroupContext(object):
+    def __init__(self, check_group_contains_valid_detectors=lambda x: True):
+        self._groups = []
+        self._runs_in_groups = []
+        self._selected = ''
+        self._selected_type = ''
+        self._selected_groups = []
+
+        self.message_notifier = MessageNotifier(self)
+
+        self._check_group_contains_valid_detectors = check_group_contains_valid_detectors
+
+    def __getitem__(self, name):
+        for item in self._groups:
+            if item.name == name:
+                return item
+        return None
+
+    @property
+    def groups(self):
+        return self._groups
+
+    @property
+    def selected_groups(self):
+        return self._selected_groups
+
+    def clear(self):
+        self._groups = []
+
+    def clear_selected_groups(self):
+        self._selected_groups = []
+
+    @property
+    def group_names(self):
+        return [group.name for group in self._groups]
+
+    def add_new_group(self, group, loadedData):
+        '''this adds groups to the grouping tab that are not already loaded'''
+        run_list = loadedData.get_parameter("run")
+        for run_item in run_list:
+            for workspace in loadedData.get_data(run=run_item)["workspace"]:
+                if str(workspace) not in self.group_names:
+                    group_name = str(workspace)
+                    '''For single workspace names the detector is found by taking everything after ; in the name
+                     For co-added workspaces the detector is found by taking everything before _'''
+                    detector_name = (group_name.split(';', 1)[-1].lstrip()).split('_', 1)[0]
+                    run_number = str(run_item).replace('[', '').replace(']', '')
+                    group += [EAGroup(group_name=group_name, detector=detector_name, run_number=run_number)]
+        return group
+
+    def add_group(self, group):
```

Comment:
whats the difference between this and add_new_group?

Reply:
I did put notes on these to try and explain but seem to have the same description on both! 

Add new group adds groups to the grouping tab if they have not already been loaded (i.e. the run has been newly added and hasn't been processed). 

On the other hand add_group adds groups to the context from the loaded data store if in fact they have already been previously loaded. This saves a bit of time, processing power and stops parts of the GUI falling over. It can't handle duplicate data in _groups as if two things have the same name it doesn't know which to pick. This prevents duplicate data being added.